lib/veri_policy.ml

```diff
@@ -0,0 +1,230 @@
+open Core_kernel.Std
+open Graph
+open Bap.Std
+open Bap_traces.Std
+open Regular.Std
+
+module Field = struct
+  type t = string [@@deriving bin_io, compare, sexp]
+  let of_string = ident
+  let empty = ""
+  let is_empty x = x = empty
+end
+
+type field = Field.t [@@deriving bin_io, compare, sexp]
+
+module Rule = struct
+  type action = Skip | Deny [@@deriving bin_io, compare, sexp]
+
+  type t = {
+    action : action;
+    insn   : field;
+    left   : field;
+    right  : field;
+  } [@@deriving bin_io, compare, fields, sexp]
+
+  let skip = Skip
+  let deny = Deny
+
+  let create ?insn ?left ?right action  = 
+    let of_opt = 
+      Option.value_map ~default:Field.empty ~f:ident in {
+      action; 
+      insn  = of_opt insn; 
+      left  = of_opt left; 
+      right = of_opt right; 
+    }
+
+  type s = t [@@deriving bin_io, compare, sexp]
+  include Regular.Make(struct
+      type t = s [@@deriving bin_io, compare, sexp]
+      let compare = compare
+      let hash = Hashtbl.hash
+      let module_name = Some "Veri_policy.Rule"
+      let version = "0.1"
+
+      let pp fmt rule = 
+        let act = match rule.action with
+          | Skip -> "Skip"
+          | Deny -> "Deny" in
+        Format.fprintf fmt "%s : %s : %s : %s ;" 
+          act rule.insn rule.left rule.right
+    end)
+end
+
+type event = Trace.event [@@deriving bin_io, sexp]
+type events = Value.Set.t
+type rule = Rule.t [@@deriving bin_io, compare, sexp]
+type trial = Pcre.regexp
+
+type matched = 
```

Comment:
A pair of list should be enough. What you're doing is trying to make a pair of list of lists. This approach will have meaning, if you were able to proof, that list are non-empty. Since, they can be empty, you introduce bogus results like, `Both (x::_),[]`, that is actually `Left` and so on